ext/msgpack/factory_class.c

```diff
@@ -54,23 +49,54 @@ static void Factory_free(msgpack_factory_t* fc)
     xfree(fc);
 }
 
-void Factory_mark(msgpack_factory_t* fc)
+void Factory_mark(void *ptr)
 {
+    msgpack_factory_t *fc = ptr;
     msgpack_packer_ext_registry_mark(&fc->pkrg);
     msgpack_unpacker_ext_registry_mark(fc->ukrg);
 }
 
-static VALUE Factory_alloc(VALUE klass)
+static size_t Factory_memsize(const void *ptr)
 {
-    msgpack_factory_t* fc = ZALLOC_N(msgpack_factory_t, 1);
+    const msgpack_factory_t *fc = ptr;
+    size_t total_size = sizeof(msgpack_factory_t);
 
-    VALUE self = Data_Wrap_Struct(klass, Factory_mark, Factory_free, fc);
-    return self;
+    if (fc->ukrg) {
+        total_size += sizeof(msgpack_unpacker_ext_registry_t) / (fc->ukrg->borrow_count + 1);
+    }
+
+    return total_size;
+}
+
+static const rb_data_type_t factory_data_type = {
+    .wrap_struct_name = "msgpack:factory",
+    .function = {
+        .dmark = Factory_mark,
+        .dfree = Factory_free,
+        .dsize = Factory_memsize,
+    },
+    .flags = RUBY_TYPED_FREE_IMMEDIATELY
+};
```

Comment:
This initialiser syntax isn't incompatible with C compilers that people are still trying to use is it?

Reply:
That's the syntax used in Ruby itself, so I assume it's fine? e.g. https://github.com/ruby/ruby/blob/46c3a93982d0b81668668a65c10bc3670a488d8a/process.c#L604-L611